text/1011-query-rewrite.rst

```diff
@@ -0,0 +1,322 @@
+
+      policy on select, update, delete
+        permit (.owner.id = global user_id)
+
+      policy on insert, update
+        check (.owner.id = global user_id)
+    }
+
+    object type Purchase extending Owned;
+
+    ...
+
+    set global user_id := <uuid-1>;
+    select count(Purchase);
+    # 9
+    set global user_id := <uuid-2>
+    select count(Purchase);
+    # 1
+
+
+Another example::
+
+    abstract object type LoggedAccess {
+      policy on select
+        do after (for obj in __result__ union (
+          insert AccessLog {
+            object := obj,
+            user_id := global user_id,
+            time := datetime_current(),
+          }
+        ))
+    }
+
+
+Bypassing policies
+==================
+
+A superuser can bypass the execution of query rewrite policies by setting
+the ``apply_rewrite_policies`` session configuration setting to ``false``.
+
+
+Introspection
+=============
+
+Policies can be introspected via a new ``schema::RewritePolicy`` in the
+introspection schema that is linked from ``schema::ObjectType`` via the new
+``policies`` link.  The ``schema::RewritePolicy`` is exposed as follows::
+
+    type schema::RewritePolicy extending schema::AnnotationSubject {
+      multi property statements -> schema::StatementKind;
+      property condition -> std::str;
+      required property action_kind -> schema::PolicyActionKind;
+      required property action -> std::str;
+    };
+
+
+Implementation considerations
+=============================
+
+Like the title of the RFC suggests, policies are a query rewrite mechanism,
+which means that they primarily affect what IR is generated for a given
+EdgeQL query.  All actions except ``DO ON COMMIT`` are pure EdgeQL transforms.
+
+``PERMIT`` and ``RESTRICT TO`` wrap set references and transform every ``Foo``
+reference into ``(SELECT Foo FILTER <permit-restrict-filter>)``.
+
+``CHECK`` actions insert an intermediate shape into ``INSERT`` and ``UPDATE``,
+e.g.::
+
+    INSERT Foo { prop := <value> }
+
+is roughly transformed into::
+
+    WITH
+      input := { prop := <value> },
+      checked := input { prop := <check_expr> }
+    INSERT Foo { prop := checked.prop }
+
+``DO AFTER`` actions transform affected statements roughly as::
+
+    WITH
+      __result__ := <orig-stmt>,
+      do_after := <do-after-expr>
+    SELECT
+      __result__
+
+When specified, the ``WHEN`` conditions must be taken into account, e.g by
+combining directly with the ``PERMIT/RESTRICT`` filters, and ``CHECK``
+expressions.  Trigger actions can be made conditional by wrapping the action
+in a ``FOR`` statement referencing the policy condition, e.g.::
+
+    FOR _ IN (SELECT true FILTER <policy-condition>)
+    UNION (<action>)
+
+``DO ON COMMIT`` is the most complex action to implement, because it requires
+two parts: 1) the query rewrite part that schedules the action by writing the
+affected data and the action expression into the job table, and 2) the runner
```

Comment:
Will job table work without too many transaction serialization errors?

Reply:
It's going to be a simple table without constraints, so hopefully we can avoid read/write dependencies between transactions on it.